Implement OAuth WebChannel key-fetch flow

rfk opened the pull request:
WIP for #2088, so that travis can watch my back as I go.  Drive-by reviews and nits most welcome at this stage, but certainly not ready for r? or merge.

shane-tomlinson commented on app/scripts/lib/fxa-client.js:
```diff
@@ -96,6 +100,8 @@ function (_, FxaClient, $, xhr, p, Session, AuthErrors, Constants) {
         updatedSessionData.unwrapBKey = accountData.unwrapBKey;
         updatedSessionData.keyFetchToken = accountData.keyFetchToken;
+      }
+      if (relier.isSync()) {
```
Good cleanup.

shane-tomlinson commented on app/scripts/lib/fxa-client.js:
```diff
@@ -24,6 +24,10 @@ function (_, FxaClient, $, xhr, p, Session, AuthErrors, Constants) {
+    // OAuth reliers can explicitly opt-in to get some derived keys.
+    if (relier.isKeyFetchEnabled()) {
+      return true;
+    }
```
The bottom two cases could be cleaned up to use the same format.

rfk commented on app/scripts/models/auth_brokers/web-channel.js:
```diff
@@ -54,6 +54,55 @@ define([
+    deriveRelierKeys: function (account) {
+      var self = this;
+      // XXX TODO: pass an fxaClient around properly
+      var fxaClient = this._assertionLibrary._fxaClient;
```
Having to access an `fxaClient` here is pretty ugly, but necessary in order to fetch the keys from `keyFetchToken`.  We could consider having `fxaClient` do the keyfetch and return `kA` and `kA` in the account dat, but this would be different to fx-desktop which wants direct access to the `keyFetchToken`.

shane-tomlinson replied:
Something about this functionality being part of an AuthBroker seems weird to me. Since it's deriveRelierKeys, can this functionality be moved to the Relier and called from here? Each AuthBroker has a reference to its relier.

rfk commented:
AFAICT this is now working more-or-less as intended.  I can point nightly at loop running inside fxa-local-dev, it will request `keys=true`, and the webchannel broker sends the appropriate keys as part of the "oauth_complete" message.

I haven't yet been able to confirm that FxAccountsOAuthClient successfully receives the keys on the other end of the webchannel; trying to log them out on the other end isn't working for some reason.  This may be some problem with my build setup.

Remaining tasks:

* Format keys as proper JWKs per https://bugzilla.mozilla.org/show_bug.cgi?id=1132293
* Figure out an official way to get an `fxaClient` inside the `getOAuthResult` call
* Resolve code-factoring questions are fxaClient.generateDerivedKey
* Documentation

shane-tomlinson commented on app/scripts/models/auth_brokers/web-channel.js:
```diff
@@ -54,6 +54,55 @@ define([
+
+    deriveRelierKeys: function (account) {
+      var self = this;
+      // XXX TODO: pass an fxaClient around properly
```
What about passing in a reference to the fxaClient to `initialize` on startup instead of peeking into the internals of the assertionLibrary?

shane-tomlinson commented on app/scripts/models/auth_brokers/web-channel.js:
```diff
@@ -54,6 +54,55 @@ define([
+      var fxaClient = this._assertionLibrary._fxaClient;
+      var masterKeys;
+      var relierKeys = {};
+      var infoPrefix = 'identity.mozilla.com/picl/v1/oauth/';
```
Can you move this prefix to [constants.js](https://github.com/mozilla/fxa-content-server/blob/14cbf41d78d918382fc414d67abc0a4fa6d81619/app/scripts/lib/constants.js)?

shane-tomlinson commented on app/scripts/models/auth_brokers/web-channel.js:
```diff
@@ -54,6 +54,55 @@ define([
+        .then(function (keys) {
+          masterKeys = keys;
+          var relierInfoA = infoPrefix + 'kAr:' + self.relier.get('clientId');
+          return fxaClient.generateDerivedKey(masterKeys.kA, 64, relierInfoA);
```
What about creating higher level functions like `generateDerivedKA` and `generateDerivedKB` that accept a `kA` or `kB` and a `clientId`? It might make the code easier to follow.

rfk commented:
@shane-tomlinson this has been updated based on your feedback and squashed into a single commit.

I moved away from the `generateDerivedKey` helper method in https://github.com/mozilla/fxa-js-client/issues/148, and instead added direct use of "sjcl" and "hkdf" into the code here.  I think it's a more sensible factoring on balance.  But it does result in duplicattion of the sjcl and hkdf code.  Any suggestions for doing this better?

vladikoff commented on bower.json:
```diff
@@ -19,6 +19,7 @@
     "requirejs-mustache": "https://github.com/jfparadis/requirejs-mustache.git#5fb8c0a3e8560526e8f9617a689e2924a8532d0a",
     "requirejs-text": "2.0.10",
     "sinon": "http://sinonjs.org/releases/sinon-1.7.1.js",
+    "sjcl": "1.0.0",
```
should we use 1.0.2?

rfk replied:
This was copypasta from fxa-js-client; certainly we should use the latest version

vladikoff commented on app/scripts/models/reliers/oauth.js:
```diff
@@ -61,6 +66,83 @@ define([
+      return Relier.prototype.isKeyFetchEnabled.call(this);
+    },
+
+    deriveRelierKeys: function (uid, keys) {
```
Nit: does it make sense to move most of this new stuff into its own module / file? and just use `deriveRelierKeys` in the model?

rfk replied:
Like a "lib/relier-keys.js" or something?

rfk commented on app/scripts/models/reliers/oauth.js:
```diff
@@ -61,6 +66,83 @@ define([
+            k: self._base64UrlEncode(keyBytes),
+            kty: 'oct',
+            rid: self.get('clientId'),
+            uid: uid
```
FYI the names `rid` and `uid` here were chosen to mirror the proposed terminolofy in https://github.com/mozilla/fxa-notification-server/blob/master/docs/overview.md#core-concepts

rfk commented:
It's not obvious where the details of this key-derivation should be documented; I added a new file in the ./docs/ folder so that they'd be officially documented *somewhere*.

rfk commented:
Updated based on @vladikoff feedback.

rfk commented:
@zaach, Shane has been out sick and may not get a chance to look at this before cutting the train.  If you could give it a look over and see what you think that would be great.

shane-tomlinson commented on app/scripts/lib/fxa-client.js:
```diff
@@ -93,9 +85,11 @@ function (_, FxaClient, $, xhr, p, Session, AuthErrors, Constants) {
       };
 
-      if (shouldFetchKeys(relier)) {
+      if (relier.isKeyFetchEnabled()) {
```
I totally dig this cleanup.

ckarlof replied:
`isKeyFetchEnabled()` is something I can imagine saying with a robotic voice. Maybe `.wantsKeys`? or `.shouldFetchKeys()`?

shane-tomlinson commented on app/scripts/lib/derived-keys.js:
```diff
@@ -0,0 +1,74 @@
+  function generateDerivedKey(key, kcls, uid, rid) {
+    key = sjcl.codec.hex.toBits(key);
+    var info = Constants.DERIVED_KEY_INFO_PREFIX + kcls + ':' + rid;
+    info = sjcl.codec.utf8String.toBits(info);
```
Two comments about lines 40:41:
* `info` is pretty vague - what sort of info is it?
* I find overwriting `info` a little bit strange, even if they are two different encodings of the same info. Would two distinct variables clarify intent?

rfk replied:
"info" is the term used in the definition of HKDF; I'll shuffle the code a bit and rename it to something like "contextInfo" which should make things a little clearer.

vladikoff replied:
`info` is used from the spec https://tools.ietf.org/html/rfc5869

```
    info     optional context and application specific information
               (can be a zero-length string)
```

shane-tomlinson commented on app/scripts/lib/fxa-client.js:
```diff
@@ -363,6 +357,13 @@ function (_, FxaClient, $, xhr, p, Session, AuthErrors, Constants) {
 
+    accountKeys: function (keyFetchToken, unwrapBKey) {
+      return this._getClient()
+              .then(function (client) {
```
nit: indentation. We started off with this formatting, then went back to the standard 2 space indent for `.then`:

```js
  return this._getClient()
    .then(function (client) {
```

shane-tomlinson commented on app/scripts/lib/hkdf.js:
```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+define(['sjcl', 'lib/promise'], function (sjcl, p) {
```
nit: we have been putting dep names on their own lines.

shane-tomlinson commented on app/scripts/lib/hkdf.js:
```diff
@@ -0,0 +1,53 @@
+   */
+  function hkdf(ikm, info, salt, length) {
+
+    var mac = new sjcl.misc.hmac(salt, sjcl.hash.sha256);
```
@rfk - who should review this portion of code to ensure correctness?

zaach replied:
This is a copy of our fxa-js-client module: https://github.com/mozilla/fxa-js-client/blob/master/client/lib/hkdf.js

ckarlof replied:
Yes, how do we DRY this without having to put this functionality in the js client? A follow on PR for this is fine.

shane-tomlinson commented on app/tests/spec/lib/fxa-client.js:
```diff
@@ -669,6 +724,23 @@ function (chai, $, sinon, FxaClient, p, testHelpers, Session, FxaClientWrapper,
+        });
+
+        return client.accountKeys('keyFetchToken', 'unwrapBKey')
+            .then(function (keys) {
```
nit: indentation

shane-tomlinson commented on app/scripts/lib/derived-keys.js:
```diff
@@ -0,0 +1,74 @@
+   * Given the account master keys, the user id and the relier id, generate
+   * the matching relier-specific derived keys.
+   */
+  function deriveRelierKeys(keys, uid, rid) {
```
When the unit tests are run with the `?coverage` query parameter, this function is not tested. It looks like the function is mocked in when needed. Can a test for this function be added?

shane-tomlinson commented on bower.json:
```diff
@@ -19,6 +19,7 @@
     "requirejs-mustache": "https://github.com/jfparadis/requirejs-mustache.git#5fb8c0a3e8560526e8f9617a689e2924a8532d0a",
     "requirejs-text": "2.0.10",
     "sinon": "http://sinonjs.org/releases/sinon-1.7.1.js",
+    "sjcl": "1.0.2",
```
Beyond the scope of this PR, but can sjcl in the fxa-js-client be exposed to eliminate the duplication?

shane-tomlinson commented:
> It's not obvious where the details of this key-derivation should be documented; I added a new file in the ./docs/ folder so that they'd be officially documented somewhere.

That looks like the correct location to me.

shane-tomlinson commented:
@rfk - aside from my nits, a missing test, and the one comment about who should review the key derivation code, looks good! Merge when comfortable.

shane-tomlinson commented:
Ah, I forgot, a functional test would be *very* handy, though not a blocker. We check to ensure the expected parameters are sent across the web channel in [tests/functional/oauth_webchannel.js](https://github.com/mozilla/fxa-content-server/blob/master/tests/functional/oauth_webchannel.js#L41)

ckarlof commented on app/scripts/lib/constants.js:
```diff
@@ -24,6 +24,9 @@ define([], function () {
     OAUTH_CODE_LENGTH: 64,
 
+    DERIVED_KEY_LENGTH: 32,
+    DERIVED_KEY_INFO_PREFIX: 'identity.mozilla.com/picl/v1/oauth/',
```
IMO, these constant names are too generic without any context.

ckarlof replied:
I could see the derived relier key module be extracted out at some point. Moving the constants to that module/file would make sense and also help clarify the context issue as well. Any other particular reason for putting them here?

shane-tomlinson replied:
> Any other particular reason for putting them here?

That was my request, more from project convention than from any hard rule. Perhaps constants.js has outlived its usefulness.

rfk replied:
I propose re-naming them to "RELIER_KEYS_XXX" in this PR for a little more clarity, and we can re-consider the overall factoring in a follow-up issue.

ckarlof commented on app/scripts/lib/derived-keys.js:
```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Likewise, this file name is too generic for my taste.

I'd prefer to talk about and name the keys about what they "are" (i.e., relier keys), rather than how they're implemented (i.e., "derived keys"). 


Talking about these as "relier keys" may make more sense

ckarlof replied:
Given the technical detail, I think more inline docs on the params of each method and their expected types would be useful. It could be a follow on PR.

ckarlof replied:
e.g., like you've done [here](https://github.com/mozilla/fxa-content-server/pull/2143/files#diff-93914b5a4e2d68ca8b53aa94239509c4R11).

ckarlof replied:
I'd like to soon see an alternate implementation of this functionality in another language to sanity check it's correctness of what it appears to be doing.

rfk replied:
Indeed.  FWIW, I hand-generated the test vectors using a python HKDF lib as a rudimentary cross-check.

rfk replied:
> I'd prefer to talk about and name the keys about what they "are" (i.e., relier keys), rather than how
> they're implemented (i.e., "derived keys"). 

Funnily enough this file started out named "relier-keys.js" and I changed it; will put it back

ckarlof commented on app/scripts/lib/derived-keys.js:
```diff
@@ -0,0 +1,74 @@
+    var relierKeys = {};
+    return generateDerivedKey(keys.kA, 'kAr', uid, rid)
+      .then(function (kAr) {
+        relierKeys.kAr = kAr;
```
Do we care that the key class name is the same as the property on the relierKeys object? Perhaps we should make 'kAr' a constant and then change this line to: `relierKeys[<constant_name>] = kAr;`. It is a little funny to have these magic strings in the code ('kAr' and 'kBr').

ckarlof commented on app/scripts/lib/derived-keys.js:
```diff
@@ -0,0 +1,74 @@
+   */
+  function deriveRelierKeys(keys, uid, rid) {
+    var relierKeys = {};
+    return generateDerivedKey(keys.kA, 'kAr', uid, rid)
```
Does this suitably blow up if `kA` or `kB` is `undefined`?

rfk replied:
I've added explicit handling and tests for this

rfk commented:
OK, rebased to address most of the nits here, thanks all.  The following remain for follow-up PRs:

* remove duplication of sjcl/hkdf libs between here and fxa-js-client
* add a functional test for the end-to-end flow
* more detailed comments on the core functions, with param types etc
* re-consider the factoring of constants etc and the possibility of extracting to a stand-alone module

ckarlof commented on app/scripts/lib/relier-keys.js:
```diff
@@ -0,0 +1,86 @@
+        if (! keys.kB) {
+          throw new Error('Cant derive relier keys: missing kB');
+        }
+        return generateDerivedKey(keys.kA, 'kAr', uid, rid);
```
`kAr` and `kBr` still seems terse and magic here. If this were factored out to a constant, it would be a more natural place to add a comment on what that string means.

rfk replied:
A function-level constant, or all the way out into constants.js?

I'm not convinced it will help much with clarity here but am happy to try.

zaach replied:
We usually lift literals to the top of the scope and give them a descriptive variable name. This is more of a nit, but `generateDerivedKey` could also take an options object instead of four arguments– the property names would provide more context on the purpose of the values.

rfk commented:
Here's an attempt at refactoring based on @ckarlof and @zaach comments.  It's not obvious what to call these "kAr" and "kBr" values, which are used both in the HKDF context info and the key id string.  I went with "key class tag" but I'm not sure that really conveys any additional information.

Maybe this smell is indicating premature generalization, and I should just make two separate methods `deriveClassAKey` and `deriveClassBKey`?

ckarlof commented on app/scripts/lib/relier-keys.js:
```diff
@@ -0,0 +1,113 @@
+  function deriveRelierKeys(keys, uid, clientId) {
+    var relierKeys = {};
+    var KEY_CLASS_TAG_A = 'kAr';
+    var KEY_CLASS_TAG_B = 'kBr';
```
I don't think that these need to be local vars. Can we hoist them up one level?

ckarlof commented on app/scripts/lib/relier-keys.js:
```diff
@@ -0,0 +1,113 @@
+   * SJCL only supports standard base64 encoding, but for proper JWKs
+   * we need the urlsafe variant.
+   */
+  function base64UrlEncode(bits) {
```
Do we have this something else already? If not can we put this in a general util lib?

ckarlof commented on app/tests/spec/lib/fxa-client.js:
```diff
@@ -107,9 +108,37 @@ function (chai, $, sinon, FxaClient, p, testHelpers, Session, FxaClientWrapper,
+          .then(function (sessionData) {
+            assert.isTrue(realClient.signUp.calledWith(trim(email), password, {
+              keys: true,
+              service: 'chronical',
```
chronicle?

ckarlof commented on app/scripts/lib/relier-keys.js:
```diff
@@ -0,0 +1,86 @@
+    var relierKeys = {};
+    return p()
+      .then(function () {
+        if (! keys.kA) {
```
An error check for missing `uid` and `rid`?

ckarlof commented:
r+ with a few minor comments to address. Looks awesome @rfk!

rfk commented:
Rebased with @ckarlof review nits addressed.  Remaining follow-ups:

* remove duplication of sjcl/hkdf libs between here and fxa-js-client
* add a functional test for the end-to-end flow
* include generation number in the JWK extra fields

rfk commented:
Follow-up issues filed, travis is green, and it works in my local testing with a suitably modified gecko.  Thanks all, let's land this thing!
